api/envoy/config/endpoint/v3/load_report.proto

```diff
@@ -159,6 +160,14 @@ message ClusterStats {
   // in the DropOverload policy.
   repeated DroppedRequests dropped_requests = 5;
 
+  // the request latencies in this cluster. Only HTTP requests are currently
+  // supported. This field is not aggregated over the load_reporting_interval
```

Comment:
This field's semantics seem to be closely tied to the Envoy implementation.  If I was implementing this in gRPC (which also uses the xDS API, including LRS), the data would be aggregated over the load reporting interval, and I would not have any idea what set of quantiles to use.

I think that in order for this API to be useful, it needs to express the data in a way that is not tied so closely to the Envoy implementation.

Reply:
Yeah, we should implement this per load reporting interval in Envoy. Working on that now.

In terms of quantiles, maybe a better API would just be:

```
// The request latencies of this cluster. This vector will have six values for:
// p25, p50, p75, p90, p95, p99
repeated double request_latency_percentiles = 7;
```

Reply:
I think those should be sufficient for most use cases

Reply:
If we are going to have the server explicitly enable collection of these stats, it could tell the client what buckets to use for the histogram.  That way, it's configurable.

Reply:
Yeah, that's a good idea. Thanks.

Reply:
Yeah, +1 on configurable. The problem though is the server might not know the client's capabilities. What are the semantics if the client only supports p50, p90 and p100, but the server wants p95? Does it try and interpolate on the histogram?

Reply:
Maybe something like this:

```
enum SupportedPercentiles {
  P25 = 1;
  P50 = 2;
  ...
}

message LoadStatsResponse {
  ...
  repeated SupportedPercentiles include_request_latency_percentiles = X;
  ...
}

message LoadStatsRequest {
  ...
  repeated SupportedPercentiles request_latency_supported_percentiles = X;
  repeated double request_latency_computed_percentiles = Y;
  ...
}
```

We have to include both supported and computed percentiles since if a server updates its requested percentiles, it needs to be able to know which ones are being sent. What do you think?

Reply:
This should also prevent users from using percentiles that aren't supported at build/compile time.

Reply:
Why would the bucket sizes need to be set at build time?  Aren't they fairly easy to compute at run-time?

Reply:
> Why would the bucket sizes need to be set at build time? Aren't they fairly easy to compute at run-time?

Do you mean quantiles? 

I think it's doable but will require some refactoring, we'll have to implement a more generic version of the [HistogramStatisticsImpl](https://github.com/envoyproxy/envoy/blob/d8b1d1191307b1890622437a75c2bd7bb9450006/source/common/stats/histogram_impl.cc#L30) since it only computes and enables a limited set of quantiles. 

Presumably we'd accept a number between 0 and 1 for the requested quantile. What should the behavior be if the requested quantile is invalid? (negative number, etc). I see that we have some kind of validator for values, maybe we can use that?

Personally it seems like a standard set of quantiles like 0.25, 0.50, 0.75, 0.95, 0.99 seems good enough, but maybe I'm missing something. I think support for this would mean I should send another PR to enable retrieving arbitrary quantiles from Envoy histograms. @htuch what do you think?

Reply:
I think some folks do care a lot about the far tail, e.g. 99.99%. I wonder if we can structure this such that:
1. The client, in the LRS request, announces the percentiles it supports.
2. The server, in the LRS response, tells the client which percentiles it is interested in (with maybe an option for a wildcard).
3. The client reports those percentiles it has available.

I actually think it's not so terrible to go back to the original idea and have the client report whatever percentiles it has on hand at some reasonable level of granularity. The server can then interpolate as needed if it needs some different bucketing.

Reply:
yeah, that sounds good. so something like:

```
enum SupportedPercentiles {
  P25 = 1;
  P50 = 2;
  ...
}

message LoadStatsResponse {
  ...
  repeated SupportedPercentiles include_request_latency_percentiles = X;
  bool send_all_request_percentiles = Y;
  ...
}

message LoadStatsRequest {
  ...
  repeated SupportedPercentiles request_latency_supported_percentiles = X;
  // these are guaranteed to be the same length
  repeated SupportedPercentiles request_latency_included_percentiles = Y;
  repeated double request_latency_computed_percentiles = Z;
  ...
}
```

Reply:
> I actually think it's not so terrible to go back to the original idea and have the client report whatever percentiles it has on hand at some reasonable level of granularity. The server can then interpolate as needed if it needs some different bucketing.

Right. We can just report all quantiles computed, but ensure that we decouple it from Envoy's internals.

Reply:
I think we could just use https://github.com/envoyproxy/envoy/blob/master/api/envoy/type/v3/percent.proto#L29 instead of `SupportedPercentile` in the request/response exchange.

In any case, I think we're converging in agreement that we need to respect the load report interval, which would be cleaner inside Envoy as well.

Reply:
Yes. Thanks for all the feedback. Working on that with @mahmoudhas